### ISSUE TITLE
assay' must be a v5 or SCT assay on ATAC integration

### ISSUE THREAD

**Issue by GuiSeSanz** [satijalab/seurat#8353]
Hello! I am working with some ATAC samples and I wanted to integrate them using the IntegrateLayers function. I am using Seurat V5 and Signac for the processing of the samples.

I create a unified set of peaks for the data to remove the artificial batch effect.
Once the data is merged, I run 

```r
seurat_merged_ATAC <- merge(
  x = H1_WT_D0_ATAC,
  y = list(H1_WT_D2_ATAC, H1_WT_D3_ATAC, H1_WT_D5_ATAC),
  add.cell.ids = c("H1_WT_D0", "H1_WT_D2", "H1_WT_D3", "H1_WT_D5")
)

seurat_merged_ATAC <- RunTFIDF(seurat_merged_ATAC)
seurat_merged_ATAC <- FindTopFeatures(seurat_merged_ATAC, min.cutoff = 20)
seurat_merged_ATAC <- RunSVD(seurat_merged_ATAC)
seurat_merged_ATAC <- RunUMAP(seurat_merged_ATAC, dims = 2:50, reduction = 'lsi')

```
Obtaining the merged samples, where I expect to see the batch effect. 
Once I try to run the integration, I get the following:
```r
integrated <- IntegrateLayers(
  object = combined, method = HarmonyIntegration,
  orig.reduction = "lsi", new.reduction = "harmony",
  verbose = TRUE
)

Error in `IntegrateLayers()`:
! 'assay' must be a v5 or SCT assay
Run `rlang::last_trace()` to see where the error occurred.


> combined
An object of class Seurat 
582832 features across 591378 samples within 1 assay 
Active assay: ATAC (582832 features, 582612 variable features)
 2 layers present: counts, data
 2 dimensional reductions calculated: lsi, umap
> class(combined)
[1] "Seurat"
attr(,"package")
[1] "SeuratObject"

```

Also, on the merged samples I get a funny behaviour... as someone with artistic aspirations has signed the umap....
![image](https://github.com/satijalab/seurat/assets/20582322/b7c48b3a-1d1a-4fdd-a0e9-7e0c3cf5e507)

```r
> sessionInfo()
R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Rocky Linux 9.1 (Blue Onyx)

Matrix products: default
BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.15.so;  LAPACK version 3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: Asia/Riyadh
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] harmony_1.2.0        Rcpp_1.0.12          future_1.33.0       
 [4] GenomicRanges_1.54.1 GenomeInfoDb_1.38.1  IRanges_2.36.0      
 [7] S4Vectors_0.40.2     BiocGenerics_0.48.1  ggrastr_1.0.2       
[10] ggplot2_3.4.4        Signac_1.12.0        Seurat_5.0.1        
[13] SeuratObject_5.0.1   sp_2.1-1            

loaded via a namespace (and not attached):
  [1] RColorBrewer_1.1-3      jsonlite_1.8.7          magrittr_2.0.3         
  [4] ggbeeswarm_0.7.2        spatstat.utils_3.0-4    farver_2.1.1           
  [7] zlibbioc_1.48.0         vctrs_0.6.5             ROCR_1.0-11            
 [10] Cairo_1.6-1             spatstat.explore_3.2-5  Rsamtools_2.18.0       
 [13] RCurl_1.98-1.13         RcppRoll_0.3.0          htmltools_0.5.7        
 [16] sctransform_0.4.1       parallelly_1.36.0       KernSmooth_2.23-20     
 [19] htmlwidgets_1.6.4       ica_1.0-3               plyr_1.8.9             
 [22] plotly_4.10.3           zoo_1.8-12              igraph_1.5.1           
 [25] mime_0.12               lifecycle_1.0.4         pkgconfig_2.0.3        
 [28] Matrix_1.6-3            R6_2.5.1                fastmap_1.1.1          
 [31] GenomeInfoDbData_1.2.11 fitdistrplus_1.1-11     shiny_1.8.0            
 [34] digest_0.6.33           colorspace_2.1-0        patchwork_1.1.3        
 [37] tensor_1.5              RSpectra_0.16-1         irlba_2.3.5.1          
 [40] labeling_0.4.3          progressr_0.14.0        fansi_1.0.6            
 [43] spatstat.sparse_3.0-3   httr_1.4.7              polyclip_1.10-6        
 [46] abind_1.4-5             compiler_4.3.0          withr_2.5.2            
 [49] BiocParallel_1.36.0     fastDummies_1.7.3       MASS_7.3-58.4          
 [52] tools_4.3.0             vipor_0.4.5             lmtest_0.9-40          
 [55] beeswarm_0.4.0          httpuv_1.6.12           future.apply_1.11.0    
 [58] goftest_1.2-3           glue_1.7.0              nlme_3.1-162           
 [61] promises_1.2.1          grid_4.3.0              Rtsne_0.16             
 [64] cluster_2.1.4           reshape2_1.4.4          generics_0.1.3         
 [67] gtable_0.3.4            spatstat.data_3.0-3     tidyr_1.3.0            
 [70] data.table_1.14.8       utf8_1.2.4              XVector_0.42.0         
 [73] spatstat.geom_3.2-7     RcppAnnoy_0.0.21        ggrepel_0.9.4          
 [76] RANN_2.6.1              pillar_1.9.0            stringr_1.5.1          
 [79] spam_2.10-0             RcppHNSW_0.5.0          later_1.3.1            
 [82] splines_4.3.0           dplyr_1.1.4             lattice_0.21-8         
 [85] survival_3.5-5          deldir_2.0-2            tidyselect_1.2.0       
 [88] Biostrings_2.70.1       miniUI_0.1.1.1          pbapply_1.7-2          
 [91] gridExtra_2.3           scattermore_1.2         matrixStats_1.1.0      
 [94] stringi_1.8.2           lazyeval_0.2.2          codetools_0.2-19       
 [97] tibble_3.2.1            cli_3.6.2               uwot_0.1.16            
[100] xtable_1.8-4            reticulate_1.34.0       munsell_0.5.0          
[103] globals_0.16.2          spatstat.random_3.2-1   png_0.1-8              
[106] parallel_4.3.0          ellipsis_0.3.2          dotCall64_1.1-0        
[109] bitops_1.0-7            listenv_0.9.0           viridisLite_0.4.2      
[112] scales_1.3.0            ggridges_0.5.4          crayon_1.5.2           
[115] leiden_0.4.3.1          purrr_1.0.2             rlang_1.1.2            
[118] fastmatch_1.1-4         cowplot_1.1.1          
```

**Comment by GuiSeSanz** [comment]
after spliting the object, the object updated from v3? to v5... 
But now, I have an additional problem and an even more cryptic one...
```r

seurat_merged_ATAC[['ATAC']] <- split(seurat_merged_ATAC[['ATAC']], f=seurat_merged_ATAC$Sample)

seurat_merged_ATAC_int <- IntegrateLayers(
  object = seurat_merged_ATAC, method = HarmonyIntegration,
  orig.reduction = "lsi", new.reduction = "harmony",
  verbose = TRUE
)
Warning: No layers found matching search pattern provided
Error in `IntegrateLayers()`:
! None of the features provided are found in this assay
Run `rlang::last_trace()` to see where the error occurred.
```

Any idea?

**Comment by mainyanghr** [comment]
I have the similar issues as well. After splitting, I can not merge or do Findmarkers due to the layers or dimetional issues.

**Comment by GuiSeSanz** [comment]
Yes, the thing is like for doing this kind of processes, the FindMarkers, the data must be unified, and you do that with `JoinLayers`. But now, the integration, instead using a list of objects, uses this splitted object...

**Comment by GuiSeSanz** [comment]
I keep trying things and each time is more evident that something is not correct....
I splitted again the object, and the results states clairly that is a V5 object....
```r
seurat_merged_ATAC_splitted <- split(seurat_merged_ATAC[['ATAC']], f=seurat_merged_ATAC$Sample)
seurat_merged_ATAC_splitted
Assay (v5) data with 582832 features for 591378 cells
Top 10 variable features:
 chr3-93470143-93471058, chr19-47256389-47257616,
chr1-15524111-15525314, chr1-6962239-6963500, chr1-16613164-16614101,
chr16-22190369-22191253, chr19-10870685-10871661,
chr19-54189402-54190356, chr9-93095396-93096377,
chr19-48492719-48493750 
Layers:
 counts.H1_WT_D0, counts.H1_WT_D2, counts.H1_WT_D3, counts.H1_WT_D5
```

But it does not work and says the same thing....

```r
seurat_merged_ATAC_int <- IntegrateLayers(
+   object = seurat_merged_ATAC_splitted, method = HarmonyIntegration,
+   orig.reduction = "lsi", new.reduction = "harmony",
+   verbose = TRUE
+ )
Error in `IntegrateLayers()`:
! 'assay' must be a v5 or SCT assay
Run `rlang::last_trace()` to see where the error occurred.
```

**Comment by mhkowalski** [comment]
Hi, 

Thanks for providing information about this issue. I'm not immediately able to reproduce this. Could you please provide a reproducible example, such as with the data used in this Signac [vignette](https://stuartlab.org/signac/articles/pbmc_vignette.html), or send us your the object you are working with? (mkowalski@nygenome.org). Thanks!

**Comment by BurhanSabuwala** [comment]
I am facing the same issue with "ADT" assay.
`merged.obj <- IntegrateLayers(merged.obj, method = HarmonyIntegration, assay = "ADT", verbose = TRUE, new.reduction = "aharmonypca", orig.reduction = "apca")
Error in `IntegrateLayers()`:
! 'assay' must be a v5 or SCT assay
Backtrace:
 1. Seurat::IntegrateLayers(...)`

**Comment by Gesmira** [comment]
Hi, 
Please let us know if you're able to provide an object. Otherwise, we will close the issue shortly.

**Comment by Nh-code** [comment]
Recently, I also encountered a similar problem in `FindMultiModalNeighbors`. It's really frustrating.
```Error in dimnames(x) <- dn : 
length of 'dimnames' [2] not equal to array extent
Calls: FindMultiModalNeighbors ... FindModalityWeights -> lapply -> FUN -> PredictAssay -> colnames<-
Execution halted```

**Comment by Gesmira** [comment]
Hi @Nh-code, this seems to be a different issue. Please open a new issue with a reproducible example, all of the code you ran that led to this error, as well as your sessionInfo(). Thanks!

**Comment by GuiSeSanz** [comment]
Sorry, I cannot share the objects as they are too heavy (16 multiome assays). 
Anyway, I solved the problem by integrating all the RNA assays together, and then adding the ATAC lsi reduction in the same cell order as I am working with multiome data.
As the integration only needs the reduced dimensionalities, it worked for me.

**Comment by kcmtest** [comment]
> Sorry, I cannot share the objects as they are too heavy (16 multiome assays). Anyway, I solved the problem by integrating all the RNA assays together, and then adding the ATAC lsi reduction in the same cell order as I am working with multiome data. As the integration only needs the reduced dimensionalities, it worked for me.

pbmc_small <- IntegrateLayers(object = pbmc_small, method = CCAIntegration, orig.reduction = "pca",
+                                 new.reduction = "integrated.cca", verbose = FALSE)
Error in `IntegrateLayers()`:
! 'assay' must be a v5 or SCT assay

any suggestion how to fix this ? since you have found a way

**Comment by bhavyaac** [comment]
> 
> pbmc_small <- IntegrateLayers(object = pbmc_small, method = CCAIntegration, orig.reduction = "pca",
> 
>     * ```
>                                   new.reduction = "integrated.cca", verbose = FALSE)
>       ```
> 
> 
> Error in `IntegrateLayers()`: ! 'assay' must be a v5 or SCT assay
> 
> any suggestion how to fix this ? since you have found a way

I solved this by setting parameter `assay` in `IntegrateLayers()` to the same as the assay being split.
For example:
```
seurat_joint[["originalexp"]] <- split(seurat_joint[["originalexp"]], f = seurat_joint$Technology)
seurat_joint_harmony <- IntegrateLayers(object = seurat_joint, 
                                              method = HarmonyIntegration, 
                                              new.reduction = "harmony", 
                                              assay = "originalexp")
    
```

**Comment by Rohit-Satyam** [comment]
I think for some reason `as.Seurat()` is creating Seurat Object with v3 assay not v5. Even using `UpdateSeuratObject` function doesn't fix this. Thankfully scCustomize has a function to fix it

```
su_rev <- lapply(su_rev, function(x){
  scCustomize::Convert_Assay(seurat_object = x, convert_to = "V5")
})
```
